### PR TITLE
 Convert Kubernetes script wrapper to not rely on scripts (bash and PS), instead do it all in c#

### PR DESCRIPTION
**Pull request by johnsimons**
This PR is to reduce the double handling of changes to both bash and Powershell scripts that should execute the same code.
I have also taken the time to add integration tests that test all supported Kubernetes authentication methods.
The tests make use of Terraform to bring up and tear down the infrastructure required to run them.

I have updated TeamCity builds to only run these new tests on 2 agents, one Linux and one Windows.

### CODE REVIEW

**Comment by johnsimons** [comment]
I have manually tested it against Azure and AWS

File: source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs
Comment (andrewabest) [inline source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs:171]: Is there a way we could break this method down a bit further? It is 300+ lines of nested statements, and is pretty tough to get your head around.

File: source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs
Comment (andrewabest) [inline source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs:21]: Awesome effort wrangling this thing out. There is a lot of behaviour in here, maintaining that in two places would have been tough going!

File: source/Calamari.Tests/KubernetesFixtures/KubernetesRunsScriptFixture.cs
Comment (andrewabest) [inline source/Calamari.Tests/KubernetesFixtures/KubernetesRunsScriptFixture.cs:70]: Should we move `SetTestClusterVariables` into `SetUp`? It is used in every test bar one.

**Review by andrewabest — APPROVED** [review]
I don't know the specifics about the correctness of the implementation, but overall looks like a good change to have made. I've made a couple of comments - the main one wondering about whether we can tame the complexity within that extended `KubernetesScriptWrapper` more. We could tackle that later though, it shouldn't block this PR going in once we are confident it is behaving correctly.

File: source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs
Comment (mcasperson) [inline source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs:204]: Errors with K8s logins have caused a few support tickets in the past. This would be a good opportunity to apply the [logging standards](https://github.com/OctopusDeploy/Architecture/blob/main/Steps/Logging.md) and link each of these errors to a unique code that links to the documentation.

File: source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs
Comment (mcasperson) [inline source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs:212]: See the comment above about linking to docs on error.

File: source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs
Comment (mcasperson) [inline source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs:225]: See the comment above about linking to docs on error.

File: source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs
Comment (mcasperson) [inline source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs:235]: See the comment above about linking to docs on error.

File: source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs
Comment (mcasperson) [inline source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs:342]: See the comment above about linking to docs on error.

File: source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs
Comment (mcasperson) [inline source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs:348]: See the comment above about linking to docs on error.

File: source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs
Comment (mcasperson) [inline source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs:375]: See the comment above about linking to docs on error.

File: source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs
Comment (mcasperson) [inline source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs:403]: See the comment above about linking to docs on error.

File: source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs
Comment (mcasperson) [inline source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs:496]: See the comment above about linking to docs on error.

File: source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs
Comment (mcasperson) [inline source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs:614]: See the comment above about linking to docs on error.

This error in particular was reported as an issue for a customer: https://octopusdeploy.slack.com/archives/C0GKZRNP4/p1621848449006000

File: source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs
Comment (mcasperson) [inline source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs:622]: https://octopusdeploy.slack.com/archives/C0GKZRNP4/p1621848449006000

File: source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs
Comment (mcasperson) [inline source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs:631]: https://octopusdeploy.slack.com/archives/C0GKZRNP4/p1621848449006000

File: source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs
Comment (mcasperson) [inline source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs:189]: This logic is going to get a lot more complex when we add gke with inherited service accounts, azure with managed service identity. It might be with extracting this if logic into a function.

**Review by mcasperson — CHANGES_REQUESTED** [review]
I like that this removes the double handling of a bash and powershell script.

I'd like to see us start using the [logging standards](https://github.com/OctopusDeploy/Architecture/blob/main/Steps/Logging.md) to direct customers to the documentation when errors are reported. We've seen a few people struggle to work out what the errors mean (this is some [customer feedback](https://octopusdeploy.slack.com/archives/C0GKZRNP4/p1621848449006000), and [internal staff have also struggled](https://octopusdeploy.slack.com/archives/C27FNL3QW/p1622151995163500), and I know of a few more issues that have come through support). By linking errors to the docs we can get the support team the ability to refine the documentation and potentially prevent support issues being passed to team steps.

**Review by paulolai — COMMENTED** [review]
I'm curious how we picked the instance sizes we did, is there a min CPU or RAM that we targeted?

File: source/Calamari.Tests/KubernetesFixtures/Terraform/Clusters/aks.tf
Comment (johnsimons) [reply source/Calamari.Tests/KubernetesFixtures/Terraform/Clusters/aks.tf:14]: These instances are only going to be up for a minute or so. So not sure there are many optimisations here.
`Standard_B2s` is not that big, we also need an instance with minimum 2 VCPUs
image

**Comment by johnsimons** [comment]
> I'd like to see us start using the [logging standards](https://github.com/OctopusDeploy/Architecture/blob/main/Steps/Logging.md) to direct customers to the documentation when errors are reported.

I'll revisit these error messages, the aim of this PR is to remove the double scripting, so I converted it 1-1.
My thinking so far on these errors @mcasperson is that we won't be seeing them that much if at all because we actually have quite comprehensive validation on when all these data is saved in the process.
So maybe we need to review what errors could really occur here.

**Comment by mcasperson** [comment]
@johnsimons 

> So maybe we need to review what errors could really occur here.

We've definitely seen permissions errors around namespaces and `kubectl` not being available on the worker as two examples that customers have identified as problems that have prevented them from completing a deployment. We won't be able to catch these errors with validation as they can change at runtime, so if these errors in particular could link to the docs that would provide a better experience.

**Comment by johnsimons** [comment]
@mcasperson I will raise Trello cards for this work

**Comment by johnsimons** [comment]
Trello card to improve errors raised in https://trello.com/c/Q20HoVyy

**Comment by johnsimons** [comment]
@mcasperson can you unblock this PR

**Pull request merged**
